Feature: Add multi databases support

spuyet opened the pull request:
The goal of this PR is to add multi database support. Non targeted migrations are still supported without requiring any changes.

Inspired by:

- https://guides.rubyonrails.org/active_record_multiple_databases.html
- https://github.com/zendesk/active_record_shards

All the examples listed below assume that the `mongoid.yml` file contains:

```yml
production:
  clients:
    default:
      database: database_one
      hosts:
        - localhost:27017
    shard42:
      database: database_two
      hosts:
        - localhost:27017
```

Here are the features added:

The migration generator behavior has been modified:

`rails generate migration` is now accepting a `--shards` parameter, the following line:

```bash
rails generate migration my_migration --shards
```

will now generate the following migration in `db/migrate/shards`:

```ruby
class MyMigration < Mongoid::Migration
  def self.up
  end

  def self.down
  end
end
```

When this migration will be executed, the new `data_migrations` entry will be stored in the `database_two` and not in the default client database.

The `db:*` tasks are now accepting a `MONGOID_CLIENT_NAME` environment variable to target a specific database:

### `db:migrate:status`

This task will show the state of the migrations that are targeted to this client **only**.

For instance, for a project with the two following migrations:

```ruby
class MyMigration < Mongoid::Migration # db/migrate/shards/20210812001200_my_migration.rb
  def self.up
  end

  def self.down
  end
end

class AnotherMigration < Mongoid::Migration # db/migrate/20210524001451_another_migration.rb
  def self.up
  end

  def self.down
  end
end
```

Here are the different outputs:

```
$ rails db:migrate:status

database: database_one

 Status   Migration ID    Migration Name
--------------------------------------------------
   up     20210524001451  AnotherMigration

$ rails db:migrate:status MONGOID_CLIENT_NAME=shard42

database: database_two

 Status   Migration ID    Migration Name
--------------------------------------------------
   up     20210812001200  MyMigration
```

### `db:drop`

```
$ rails db:drop
```

will delete `database_one`

```
$ rails db:drop MONGOID_CLIENT_NAME=shard42
```

will delete `database_two`

### `db:version`

```
$ rails db:version
```

will display `20210524001451`

```
$ rails db:version MONGOID_CLIENT_NAME=shard42
```

will display `20210812001200`

### `db:migrate`

This task executes all down migrations that are not using a custom client and write the `data_migrations` entries in the default client database, `database_one` in our example.

When the `MONGOID_CLIENT_NAME` environment variable is specified, the task will migrate all the down migrations in the `shards` folder, the `data_migrations` entries will be added in the database linked to the client passed as parameter, `database_two` in our example.

The following command:

```
  rails db:migrate
```

will only execute `AnotherMigration` migration, while:

```
  rails db:migrate MONGOID_CLIENT_NAME=shard42
```

will only execute `MyMigration`.

### `db:migrate:down`

This task will revert the migration specified by the `VERSION` environment variable if it belongs to the current target, then the `data_migrations` entry will be removed from the database linked to this migration.

The following command:

```
rails db:migrate:down VERSION=20210524001451
```

will revert the migration and remove the `20210524001451` entry from the `data_migrations` collection for the `default` client.

```
rails db:migrate:down VERSION=20210524001451 MONGOID_CLIENT_NAME=shard42
```

will raise a `UnknownMigrationVersionError` as this migration is not tagged for this client.

```
rails db:migrate:down VERSION=20210812001200 MONGOID_CLIENT_NAME=shard42
```

will revert the migration and remove the `20210812001200` entry from the `data_migrations` collection for the `shard42` client.

### `db:migrate:up`

This task will execute the migration specified by the `VERSION` environment variable if it belongs to the current target, then the `data_migrations` entry will be added to the database linked to this migration.

The following command:

```
rails db:migrate:up VERSION=20210524001451
```

will execute the migration and add the `20210524001451` entry to the `data_migrations` collection for the `default` client.

```
rails db:migrate:up VERSION=20210524001451 MONGOID_CLIENT_NAME=shard42
```

will raise a `UnknownMigrationVersionError` as this migration is not tagged for this client.

```
rails db:migrate:up VERSION=20210812001200 MONGOID_CLIENT_NAME=shard42
```

will execute the migration and add the `20210812001200` entry to the `data_migrations` collection for the `shard42` client.

### `db:migrate:redo`

When the `VERSION` environment variable is specified, this task calls `db:migrate:down` followed by `db:migrate:up`. Otherwise, it calls `db:rollback` and `db:migrate`.

The `MONGOID_CLIENT_NAME` environment variable is needed to execute a migration that is linked to a specific client.

### `db:rollback`

If the `STEP` environment variable is defined, it will rollback the last `STEP` migrations on the default database (=> `database_one`).

Otherwise, it rollbacks only the last migration stored in `database_one`.

When the `MONGOID_CLIENT_NAME` environment variable is defined, it will rollback the migrations on the targeted database only, meaning that for the following command:

```
rails db:rollback STEP=3 MONGOID_CLIENT_NAME=shard42
```

The last 3 migrations that have been executed on `database_two` will be rollbacked and removed from `database_two`.

### `db:rollback_to`

This task will revert all the migration created since the migration specified by the `VERSION` environment variable and that are not defined to run on a shard client.

When the `MONGOID_CLIENT_NAME` environment variable is defined, it will rollback all the migrations that are up for this client between the migration specified and now.

spuyet commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -216,7 +236,10 @@ def rollback(migrations_path, steps=1)
         all_versions = get_all_versions
-        rollback_to = all_versions.index(target_version.to_i) + 1
+        target_version_index = all_versions.index(target_version.to_i)
+        raise UnknownMigrationVersionError.new(target_version) if target_version_index.nil?
```
This is fixing a previous bug, when a wrong version was used in:
`rails db:rollbkack_to VERSION=fake_version`
this code was raising the following error:
```
rails aborted!
NoMethodError: undefined method `+' for nil:NilClass
/home/seb/projects/mongoid_rails_migrations/lib/mongoid_rails_migrations/active_record_ext/migrations.rb:239:in `rollback_to'
/home/seb/projects/mongoid_rails_migrations/lib/mongoid_rails_migrations/mongoid_ext/railties/database.rake:113:in `block (3 levels) in <main>'
/home/seb/projects/mongoid_rails_migrations/lib/mongoid_rails_migrations/active_record_ext/migrations.rb:294:in `with_mongoid_client'
/home/seb/projects/mongoid_rails_migrations/lib/mongoid_rails_migrations/mongoid_ext/railties/database.rake:112:in `block (2 levels) in <main>'
/home/seb/ringcentral/engage-digital/bin/rails:8:in `<top (required)>'
/home/seb/ringcentral/engage-digital/bin/spring:16:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'
Tasks: TOP => db:rollback_to
(See full trace by running task with --trace)
```

spuyet commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -430,10 +474,10 @@ def record_version_state_after_migrating(version)
         else
           @migrated_versions.push(version).sort!
-          DataMigration.create(:version => version.to_s)
+          DataMigration.find_or_create_by(:version => version.to_s)
```
This change and the one line `477` prevent multiples entries with the same version to be written in the `data_migrations` collection.

spuyet replied:
FWICS in code yes but I found in one of our database that sometimes it seems not working as expected:
```
> DataMigration.where(version: "20170809160116").count
=> 6
```
Not sure of why, the migration probably failed multiple times but when it failed, it should haven't added multiple entries in the `data_migrations` collection.

As the same version cannot be used as multiple migrations, we're safe to add this IMO.

jarthod replied:
Why not ^^

spuyet commented on test/helper.rb:
```diff
@@ -9,11 +9,30 @@
+end
+
+def invoke(task)
+  Rake.application.tasks.each(&:reenable)
```
We can't execute the same rake task multiple times without this, that's sad but default `Rake` behavior.

jarthod replied:
And can't we reenable just the task? like `Rake::Task[task].reenable`?

spuyet replied:
Not really, some tasks are calling other tasks and reenabling only the called task does not reenable the ones called by the caller.

jarthod replied:
Ok

spuyet commented:
Integration seems to fail because of:
```
File does not exist: minitest/metametameta
```
Not sure why, specs work well locally:

```
$ bundle exec rake
Run options: --seed 25949

# Running:

...........................................................

Finished in 2.425746s, 24.3224 runs/s, 71.3183 assertions/s.
59 runs, 173 assertions, 0 failures, 0 errors, 0 skips
```

FabienChaynes commented on lib/mongoid_rails_migrations/mongoid_ext/railties/database.rake:
```diff
@@ -1,17 +1,23 @@
-    desc 'Drops all the collections for the database for the current Rails.env'
-    task :drop => :environment do
-      Mongoid.master.collections.each {|col| col.drop_indexes && col.drop unless ['system.indexes', 'system.users'].include?(col.name) }
+  if Rake::Task.task_defined?("db:drop")
```
Why this change?

spuyet replied:
To provide a way to drop a specific database.
Also the `Mongoid.master` method does not exist anymore.

jarthod replied:
Mm but why would we need to `clear` existing task more than before?

Also to get the client database you can use `Mongoid.current_client.database` instead of going through the DataMigration collection.

spuyet replied:
> Mm but why would we need to clear existing task more than before?
If we don't do this, it will append this task to the existing one, we can keep it but it could drop the database twice what is particularly useless.

> Also to get the client database you can use Mongoid.current_client.database instead of going through the DataMigration collection.

Good point :+1:

jarthod replied:
Which existing one? from Rails? the gem was not doing this before so it was not working and using ActiveRecord task?

spuyet replied:
I don't know, probably the one defined by `ActiveRecord` yes, you can see in the diff that before we were defining the task only if it was not defined:
```ruby
unless Rake::Task.task_defined?("db:drop")
  desc 'Drops all the collections for the database for the current Rails.env'
  task :drop => :environment do
    Mongoid.master.collections.each {|col| col.drop_indexes && col.drop unless ['system.indexes', 'system.users'].include?(col.name) }
  end
end


```

jarthod replied:
Yes so the question still stands: why did you change this if it was like this before?

spuyet replied:
I've already answered to this question:
> If we don't do this, it will append this task to the existing one, we can keep it but it could drop the database twice what is particularly useless.

jarthod replied:
Maybe I'm missing something but it still does not answer my question: why do we need this NOW and not BEFORE?

spuyet replied:
I think I've already answered to this question above but I will reformulate with more deails:

**BEFORE**
The task was already present in the mongoid gem:
```ruby
desc "Drops the default client database"
  task :drop => :environment do
    ::Mongoid::Clients.default.database.drop
end
```
This is deleting the database for the default client **ONLY**.

**NOW**
We're adding multi shards so we propose to pass the targeted database as parameter (in the env variable) like:
```
MONGOID_CLIENT_NAME=shard1 rails db:drop
```
We are now removing the old task **BECAUSE** if we don't this both task will be called and that's not what we want. It was not necessary before the default task was already doing the expected job.

jarthod replied:
Ok thanks, that's better indeed :+1:
So this means the previous task was useless though (as already defined in Mongoid)? Or maybe it was for very old versions of mongoid which did not define it?

spuyet replied:
Most probably yes, FWICS the database management tasks appeared only in 2014 in the mongoid gem: https://github.com/mongodb/mongoid/commits/master/lib/mongoid/tasks/database.rake before it was using the ActiveRecord task while this gem added support database drop from the beginning in 2010: https://github.com/adacosta/mongoid_rails_migrations/commit/39e5004666cf895a8fe15051a76fcd6828266740#diff-b87e05cc243468dabc4100dbfdd30f248ac332282dca6bb7655eef9a0ee303aa

spuyet commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -352,6 +395,7 @@ def status
       puts "-" * 50
       up_migrations = migrated.to_set
       migrations.each do |migration|
+        next unless client_match?(migration)
```
Yep, that's true, now it loads the migration file to detect the client and for migrations loading files that does not exist anymore it will fail.

My POV on this is that we should not keep hundreds migration files on a project like in this example:
```
$ RAILS_ENV=staging bundle exec rails db:migrate:status | grep up | wc -l
644
```
I try to keep only the last 3 years of migrations files on the same project, it works well and still contains a lot of migration files:
```
$ rails db:migrate:status | grep up | wc -l
157
```
Moreover, keeping legacy code in codebase is misleading IMO and can lead newcomers to use deprecated methods in their own migrations.

> It could maybe be handled by grouping migrations for the same non-default client in a subfolder? This way it'd be organized and we could infer the client thanks to the folder name without loading the migration?

That's a good idea but I fear that that could be a bit confusing for users in the end as we've already [multi folders support ](https://github.com/adacosta/mongoid_rails_migrations/blob/master/lib/mongoid_rails_migrations/active_record_ext/migrations.rb#L362) that is not only doing more than merging all migrations in the same array.

I'm not saying no, just explaining why I choose this opinionated way ;)

WDYT @jarthod ?

jarthod replied:
Both good points, quick question: does it fail on our project? because it's true we now need to load the migration class but we don't need to run the `up` or `down` which are the ones containing the legacy code usually. So unless there's some deprecated code used in the migration class directly (like to build a constant), this has low change to fail.

Anyway we can discuss this again after we address the bigger concern of my review.

spuyet replied:
Yes, it fails for old migrations requiring files that are not present in the project anymore. As I explained above, it does not fail for 3 years old migrations but for 10 years old yes.

jarthod replied:
Ok thanks, for our project I don't care much and I agree we could just remove the old or failing migrations, but what I don't like is that this is gonna fail for anybody using the gem, even if they don't use sharding (nobody will). So I would say it's ok to keep it like this IF you change the code to only load the migration when using MONGOID_CLIENT_NAME, does that sound possible? so for 99.9% of users this should not change anything.

spuyet replied:
That's a good idea but seems not feasible as when we do not use `MONGOID_CLIENT_NAME`, we need to know if the migration targets the default client or not, if not we've to skip the migration. If you think that's a pain, we can go for the folder way.

We could also require to enable a `sharding` option in config in order to make the client check only when the `sharding` is turned on.

jarthod replied:
Mmm indeed, but maybe we can totally ignore this when MONGOID_CLIENT_NAME is not passed? Which means force people to always use "MONGOID_CLIENT_NAME" when using sharding? (maybe a bit dangerous). Otherwise I guess the folders seems best indeed.

spuyet commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -296,19 +331,25 @@ def current_migration
-      unless (up? && migrated.include?(target.version.to_i)) || (down? && !migrated.include?(target.version.to_i))
-        target.migrate(@direction)
-        record_version_state_after_migrating(target.version)
+      return unless client_match?(target)
```
That's a good point :+1:

spuyet replied:
This logic changed and now if migration does not exist for the client passed, it will raise a `UnknownMigrationVersionError`.

jarthod commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -268,6 +291,14 @@ def proper_table_name(name)
 
+      def with_mongoid_client(mongoid_client_name, &block)
+        previous_mongoid_client_name = Thread.current[:mongoid_client_name]
+        Thread.current[:mongoid_client_name] = mongoid_client_name
```
Using a Thread variable to share and leak the client name is quite intrusive IMO, it works well in our case because we use the same (and with the same value) but it could be very confusing for people using a different variable or sharding implementation (as this gem does not provide the sharding implementation).

I think here we can more simply set the mongo client around each migration without leaking a thread variable, can't we? we don't need this level of precision which is actually more for sharding an application. Here there's only one model and the rest is up to the application.

spuyet replied:
> Using a Thread variable to share and leak the client name is quite intrusive IMO, it works well in our case because we use the same (and with the same value) but it could be very confusing for people using a different variable or sharding implementation (as this gem does not provide the sharding implementation).

Yep but how do we handle the `db:migrate:status`, `db:version` tasks to target a particular client ? Using an environment variable ? That's what I did first but in the end it seemed more harmonious to me to use the same way of switching client in the whole codebase.

> I think here we can more simply set the mongo client around each migration without leaking a thread variable, can't we? we don't need this level of precision which is actually more for sharding an application. Here there's only one model and the rest is up to the application.

Not sure of what you want to say by `around each migration`, the client needs to be changed only when we add the migration entry in the `data_migrations` collection not during the migration execution.

We could replace this method with `Mongoid.with_client` around the code that add/remove entries in `data_migrations` if it's what you asked, yes.

jarthod replied:
> Yep but how do we handle the db:migrate:status, db:version tasks to target a particular client ? Using an environment variable ? That's what I did first but in the end it seemed more harmonious to me to use the same way of switching client in the whole codebase.

What's the problem with these tasks?

> Not sure of what you want to say by around each migration, the client needs to be changed only when we add the migration entry in the data_migrations collection not during the migration execution.
>
> We could replace this method with Mongoid.with_client around the code that add/remove entries in data_migrations if it's what you asked, yes.

I think the standard way to change the mongoid client is with `Mongoid.override_client(name)` now, and we can save/restore the previous one to be clean with something like https://github.com/pavlo/mongoid-dynamic-clients/blob/develop/lib/mongoid/dynamic_clients/client_switch.rb. `with_client` is a monkey-patch from us from way before Mongoid implemented `override_client`.

But yes that's what I mean. And yes theoretically we could only do this for the DataMigration call, but it would be weird and error prone IMO to have a migration with "client :shard" execute with the default client by default don't you think? In our case we don't care I think because our project will change the client anyway but for other users of the gem it may be more logical to have the client designated in the migration during the whole migration execution.

jarthod commented on lib/mongoid_rails_migrations/mongoid_ext/railties/database.rake:
```diff
@@ -33,24 +39,30 @@ namespace :db do
 
   desc 'Current database version'
   task :version => :environment do
```
Putting it here but it applies to others: most of these tasks load the whole rails `environment` but I don't think they all actually need it? if we can remove some of the we can make them a lot faster (not related to this MR of course, could just be a nice side improvement IF It does not require more changes)

spuyet replied:
Good point, let's create a dedicated issue for this as this is not related to this MR.

jarthod commented on lib/rails/generators/mongoid/migration/templates/migration.rb:
```diff
@@ -1,5 +1,5 @@
 class <%= migration_class_name %> < Mongoid::Migration
-  def self.up
+<%= "  client :#{client_name}\n\n" if client_name %>  def self.up
```
won't this put the `def self.up` on the same line as the client?

spuyet replied:
Nope, there is two `\n` in the generated string. Anyway, I double checked and adding `-` at the end seems to remove the trailing carriage return when the string returned by erb is empty, so the template will be easier to read like this:
```ruby
  <%= "client :#{client_name}\n" if client_name -%>
  def self.up
```

jarthod replied:
Ah ok good point, but yes indeed it looks nicer with the `-%>`. Thanks!

jarthod reviewed (changes requested):
:warning: My main problem about this MR is this `client :shard1` parameter, how can you run one migration on shard1, shard2 and then shard3 with this syntax? That's why `active_record_shard` gem has a `shard :all` syntax. Maybe I missed something here? AFAICS this can't work. We would need some kind of regexp or prefix matching maybe, or reverse and specify only when it's the default client and otherwise default to the current one set by the ENV variable. (but if we do that it's a bit tricky to prevent shard migration to be ran on the default client too)

spuyet commented:
>  My main problem about this MR is this `client :shard1` parameter, how can you run one migration on shard1, shard2 and then shard3 with this syntax? That's why `active_record_shard` gem has a `shard :all` syntax. Maybe I missed something here? AFAICS this can't work. We would need some kind of regexp or prefix matching maybe, or reverse and specify only when it's the default client and otherwise default to the current one set by the ENV variable. (but if we do that it's a bit tricky to prevent shard migration to be ran on the default client too)

Got your point but that's not that simple in our case. In theory, you want to be able to run the migration on shared database OR on the shards meaning that for a sharded migration it will run the migration for all shards, but if you want to run a migration for only one shard among the list of clients that you have in your `mongoid.yml` file ? How do you proceed ?

Here is how I saw that could work in our context:
- Add a new `shard` client to all servers pointing to the current shard mongos
- Generate all sharded mongo migrations with the `shard` client

We could tweak the `mongoid.yml` file to add a new `shards` key that could contains the only client that we want to target on every server, why not.

jarthod commented:
> Here is how I saw that could work in our context:
> 
> * Add a new `shard` client to all servers pointing to the current shard mongos
> * Generate all sharded mongo migrations with the `shard` client
> 
> We could tweak the `mongoid.yml` file to add a new `shards` key that could contains the only client that we want to target on every server, why not.

That's basically avoiding the problem here and moving it to the application which makes it ever more complicated for anybody else to use this feature and means we'll always live with duplicated clients in mongoid.yml forever. Not a good enough option IMO. Also this will not work locally and on staging for example.

> for only one shard among the list of clients that you have in your mongoid.yml file

Why would you ever want that? shards share the same code we don't want different data there. And if some migration is only important for one shard because it's data-related only, we should make it run and do nothing on other shards so we have a consistent list of migration on all shards IMO.

spuyet commented:
>  That's basically avoiding the problem here and moving it to the application which makes it ever more complicated for anybody else to use this feature and means we'll always live with duplicated clients in mongoid.yml forever. Not a good enough option IMO. Also this will not work locally and on staging for example.
> Why would you ever want that? shards share the same code we don't want different data there.

That's a workaround yes, if we want to have something like migrations on all shards OR shared database, we need to remove the clients from other shards on servers and keep the client for the current shard only, it means removing the global stuff, like for example global billing worker that we're not able to delete because we want to generate only one billing file and not one per shard. FTR, we're also running against the clock here.

We also have replica client BTW in `mongoid.yml`, how do we deal with these clients ?

> And if some migration is only important for one shard because it's data-related only, we should make it run and do nothing on other shards so we have a consistent list of migration on all shards IMO.

So it means adding support of migration to a targeted shard, that's not a migration for all shards OR only for the shared database.

spuyet commented:
Ok, I took time to think more about this and indeed I'm trying to kill two birds with one stone. Our context is very specific and the changes with migrations for all shards OR only shared are not really a blocker, we will just continue to run all the migrations from one server only, but we can improve this later on our side.

So here my proposal:
- Migration will be executable for all shards or shared db only, no support of migrations for a particular shard
- Use the client with name starting by shard as shard members (we will have rename our replication mongoid clients from `shard1_secondary` to `secondary_shard1` for example. If we are too strict on client name format, it could break some custom shard client name like `shard1-us1`.
- What is weird in this way is that for a sharded migration, we're going to run the migration only once but add the migration entry to all sharded databases. I agree that we're not providing models sharding but that's still a bit weird.

jarthod commented:
{quote}
So it means adding support of migration to a targeted shard, that's not a migration for all shards OR only for the shared database
{quote}
Well that's what you were saying we should support isn't it? and on the other hand I said we should not do anything specific here. If the project needs to run a migration on shard1 only, it can add a `return if` in the migration directly, this should not be the responsibility of the migration gem IMO.

{quote}
Use the client with name starting by shard as shard members (we will have rename our replication mongoid clients from shard1_secondary to secondary_shard1 for example. If we are too strict on client name format, it could break some custom shard client name like shard1-us1.
{quote}
I'm not sure why the secondary client would be a problem? as we'll run the migrations with MONGOID_CLIENT_NAME=shard1 anyway? We just need to make sure the migration can support multiple different "MONGOID_CLIENT_NAME", but other than that the code will simply use the requested client, it does not need to iterate or find all the matching clients does it?

{quote}
What is weird in this way is that for a sharded migration, we're going to run the migration only once but add the migration entry to all sharded databases. I agree that we're not providing models sharding but that's still a bit weird
{quote}
I'm not sure to understand this one, are you talking about a migration which applies only to a single shard? if so like I said above this should not be the responsibility of the gem IMO, because it's very rare and there can be special cases (like execute on shard1 and 2 but not shard3). That's why I think it's better to leave this to the application and just allow the migration to run on all shards from the gem point of view.

spuyet commented:
Ok, following the previous comment, I've updated the logic, the MR description has been updated too.
The migrations for shards can now be created with the `--shards` option and will be located in the `shards` subfolder. Migration are also runnable for all shards, not for a specific shard.

spuyet commented on mongoid_rails_migrations.gemspec:
```diff
@@ -19,7 +19,7 @@ Gem::Specification.new do |spec|
 
   spec.add_runtime_dependency('bundler', '>= 1.0.0')
-  spec.add_runtime_dependency('mongoid', '>= 4.0.0')
+  spec.add_runtime_dependency('mongoid', '>= 5.0.0')
```
Mongoid >= 5.0 is needed to use `Mongoid::Threaded`

spuyet commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -153,10 +153,12 @@ def suppress_messages
+            Mongoid.client(Mongoid::Threaded.client_override)
+          end
         else
-          ::Mongoid.default_session
```
This method does not exist anymore in Mongoid >= 5.0

jarthod commented on README.md:
````diff
@@ -12,6 +12,11 @@ Create migration
+```
+$ rails generate mongoid:migration <your_migration_name_here> --shards
+```
+
````
That's a good start but we still need some explanation about what is sharding in this gem and how it works IMO

jarthod commented on lib/mongoid_rails_migrations/active_record_ext/migrations.rb:
```diff
@@ -268,6 +277,14 @@ def proper_table_name(name)
+        block.call
+      ensure
+        Mongoid.override_client(previous_mongoid_client_name)
+      end
```
That looks cleaner, thanks :+1:

jarthod reviewed (approved):
Nice, looks better to me, WDYT @FabienChaynes ?
I think there's still some work on the readme/documentation though :)

FabienChaynes commented on test/helper.rb:
```diff
@@ -9,11 +9,30 @@
+end
+
+def with_env(options)
+  options.keys.each { |key| ENV[key] = options[key] }
```
```suggestion
  options.each { |k, v| ENV[k] = v }
```

Really minor but cleaner.

FabienChaynes commented on test/tasks/down_task_test.rb:
```diff
@@ -0,0 +1,157 @@
+   up     20210210125532  ShardDatabaseMigrationTwo
+EOF
+
+      with_env("MONGOID_CLIENT_NAME" => "shard1", "VERSION" => "20210210124656") do
```
```suggestion
      with_env("MONGOID_CLIENT_NAME" => "shard1") do
```

Not used for `db:migrate:status`. Same for the other places like this in the specs.

FabienChaynes commented on test/tasks/drop_task_test.rb:
```diff
@@ -0,0 +1,43 @@
+      invoke("db:drop")
+      assert_output("0\n") { invoke("db:version") }
+      with_env("MONGOID_CLIENT_NAME" => "shard1") do
+        invoke("db:migrate")
```
Minor, but what's the goal of the `db:migrate` here? Isn't just the `db:version` enough to make sure that it wasn't dropped?

FabienChaynes commented on test/tasks/migrate_task_test.rb:
```diff
@@ -0,0 +1,166 @@
+      invoke("db:migrate")
+      with_env("MONGOID_CLIENT_NAME" => "shard1") do
+        assert_output(output) { invoke("db:migrate:status") }
+        output = <<-EOF
```
You can reuse the previous output if you want.

FabienChaynes commented on test/tasks/redo_task_test.rb:
```diff
@@ -0,0 +1,152 @@
+   up     20100513055502  AddSecondSurveySchema
+   up     20100513063902  AddImprovementPlanSurveySchema
+EOF
+      ) { invoke("db:migrate:status") }
```
There was no `down` migration before the redo, so there's no proof that the redo with the version is not targeting the last version (same for the next specs).

FabienChaynes commented on test/tasks/redo_task_test.rb:
```diff
@@ -0,0 +1,152 @@
+   up     20210210124656  ShardDatabaseMigration
+   up     20210210125532  ShardDatabaseMigrationTwo
+EOF
+        ) { invoke("db:migrate:status") }
```
I'd keep it `down` here to be sure that the redo is on the non sharded database. Same for the next spec.

FabienChaynes commented on test/tasks/reset_task_test.rb:
```diff
@@ -0,0 +1,124 @@
+   up     20100513063902  AddImprovementPlanSurveySchema
+EOF
+      ) { invoke("db:migrate:status") }
+      invoke("db:migrate:reset")
```
Same here, I think it's better to have it `down` (or at least partially `down`) to begin with in order to be sure that something was done.

FabienChaynes commented on test/tasks/up_task_test.rb:
```diff
@@ -0,0 +1,141 @@
+  down    20100513063902  AddImprovementPlanSurveySchema
+EOF
+      ) { invoke("db:migrate:status") }
+      ENV["VERSION"] = "20100513055502"
```
We could use `with_env`.

FabienChaynes reviewed (commented):
Good to me on the "real" changes, most comments are about specs (which are really complete BTW :+1:).

spuyet commented:
@jarthod @FabienChaynes I've applied comment about tests in : https://github.com/adacosta/mongoid_rails_migrations/pull/57/commits/fb2c281a4d1a3f30ed65d73020050033cfb189d8 and added details about multi databases support in readme https://github.com/adacosta/mongoid_rails_migrations/pull/57/commits/72758337316563fa6851c45e78a6a99679e6228d


(I force pushed because I rebased the commit about tests).

jarthod commented on README.md:
````diff
@@ -26,20 +26,45 @@ $ rails db:reseed (handled by mongoid)
+$ rails db:migrate MONGOID_CLIENT_NAME=shard42
+```
+
+The shards migrations will be executed and entries will be added in the `shard42` database.
````
Thanks for this new section, it's much better :)
To be precise, it's not the shard42 "database" but the mongodb (hostname+port+database) referenced by the "shard42" client in mongoid.yml. Maybe you could also add a small mongoid.yml example with default + 2 shards to make this clearer.

spuyet replied:
Good point, I didn't want to make this section too big but you're right that will be clearer with an example

spuyet replied:
@jarthod I've updated the readme here: https://github.com/adacosta/mongoid_rails_migrations/pull/57/commits/8cc5f8a4c355eeb8b68e6072ac4e42fac4178c7f

jarthod replied:
Look good, thanks!

jarthod reviewed (commented):
Little comment about the readme but otherwise looks good to me!

FabienChaynes commented on test/tasks/up_task_test.rb:
```diff
@@ -0,0 +1,141 @@
+  down    20210210125532  ShardDatabaseMigrationTwo
+EOF
+
+      with_env("MONGOID_CLIENT_NAME" => "shard1", "VERSION" => "20210210124656") do
```
```suggestion
      with_env("MONGOID_CLIENT_NAME" => "shard1") do
```

FabienChaynes commented on test/tasks/up_task_test.rb:
```diff
@@ -0,0 +1,141 @@
+      with_env("VERSION" => "20210210125000") do
+        invoke("db:migrate:up")
+      end
+      with_env("MONGOID_CLIENT_NAME" => "shard1", "VERSION" => "20210210124656") do
```
```suggestion
      with_env("MONGOID_CLIENT_NAME" => "shard1") do
```

FabienChaynes commented on test/tasks/drop_task_test.rb:
```diff
@@ -0,0 +1,39 @@
+        assert_output("20210210125532\n") { invoke("db:version") }
+      end
+      invoke("db:drop")
+      assert_output("0\n") { invoke("db:version") }
```
```suggestion
      assert_output("0\n") { invoke("db:version") }
      with_env("MONGOID_CLIENT_NAME" => "shard1") { assert_output("20210210125532\n") { invoke("db:version") } }
```

To be sure that it does not drop the sharded db (apparently my previous comment wasn't clear. I just wanted to remove the `db:migrate` for shard1, not the `db:version` test after the drop).

FabienChaynes reviewed (changes requested):
Miscommunication about a previous comment + 2 useless `VERSION` env vars for `db:migrate:status` but nothing critical.

jarthod commented:
Thanks, this is now merged into master though I would like to have some feedback before cutting a release so could you please update to the master branch for now?
